src/responder/common/cache_req/cache_req.c

```diff
@@ -0,0 +1,558 @@
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <ldb.h>
+#include <talloc.h>
+#include <tevent.h>
+
+#include "util/util.h"
+#include "responder/common/cache_req/cache_req_private.h"
+#include "responder/common/cache_req/cache_req_plugin.h"
+
+static struct cache_req_plugin *
+cache_req_get_plugin(enum cache_req_type type)
+{
+    static struct cache_req_plugin *plugins[CACHE_REQ_SENTINEL] = {
+
+    };
+
+    if (type >= CACHE_REQ_SENTINEL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Bug: invalid plugin type!");
+        return NULL;
+    }
+
+    return plugins[type];
+}
+
+static void cache_req_set_plugin(struct cache_req *cr,
+                                 enum cache_req_type type)
+{
+    struct cache_req_plugin *plugin;
+
+    plugin = cache_req_get_plugin(type);
+    if (plugin == NULL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Bug: unset plugin!");
+        return;
+    }
+
+    cr->reqname = plugin->name;
+    cr->dp_type = plugin->dp_type;
+    cr->plugin = plugin;
+
+    CACHE_REQ_DEBUG(SSSDBG_TRACE_INTERNAL, cr, "Setting \"%s\" plugin\n",
+                    plugin->name);
+
+    return;
+}
+
+static struct cache_req *
+cache_req_create(TALLOC_CTX *mem_ctx,
+                 struct resp_ctx *rctx,
+                 struct cache_req_data *data,
+                 struct sss_nc_ctx *ncache,
+                 int midpoint)
+{
+    struct cache_req *cr;
+
+    cr = talloc_zero(mem_ctx, struct cache_req);
+    if (cr == NULL) {
+        return NULL;
+    }
+
+    cr->rctx = rctx;
+    cr->data = data;
+    cr->ncache = ncache;
+    cr->midpoint = midpoint;
+    cr->req_start = time(NULL);
+
+    /* It is perfectly fine to just overflow here. */
+    cr->reqid = rctx->cache_req_num++;
+
+    cache_req_set_plugin(cr, data->type);
+
+    return cr;
+}
+
+static errno_t
+cache_req_set_name(struct cache_req *cr, const char *name)
+{
+    const char *dup_name;
+
+    CACHE_REQ_DEBUG(SSSDBG_TRACE_FUNC, cr, "Setting name [%s]\n", name);
+
+    dup_name = talloc_strdup(cr->data, name);
+    if (dup_name == NULL) {
+        CACHE_REQ_DEBUG(SSSDBG_CRIT_FAILURE, cr, "Unable to set name!\n");
+        return ENOMEM;
+    }
+
+    talloc_zfree(cr->data->name.name);
+    cr->data->name.name = dup_name;
+
+    return EOK;
+}
+
+static errno_t
+cache_req_prepare_domain_data(struct cache_req *cr,
+                              struct sss_domain_info *domain)
+{
+    errno_t ret;
+
+    if (cr->plugin->prepare_domain_data_fn == NULL) {
+        return EOK;
+    }
+
+    CACHE_REQ_DEBUG(SSSDBG_TRACE_FUNC, cr,
+                    "Preparing input data for domain [%s] rules\n",
+                    domain->name);
+
+    ret = cr->plugin->prepare_domain_data_fn(cr, cr->data, domain);
+    if (ret != EOK) {
+        CACHE_REQ_DEBUG(SSSDBG_CRIT_FAILURE, cr,
+                        "Unable to prepare data [%d]: %s\n",
+                        ret, sss_strerror(ret));
+        return ret;
```

Comment:
I am afraid I am not fine with this ```return``` and another one three lines below. It could sound like nitpick but I prefer to fix it.